examples/stabilizer_code.py

```diff
@@ -0,0 +1,340 @@
+    for r in reversed(range(rank)):
+        i = min_row + r
+        j = min_col + r
+
+        # Do the elimination.
+        for k in reversed(range(min_row, i)):
+            if M[k, j] == 1:
+                M[k, :] = np.mod(M[i, :] + M[k, :], 2)
+
+    return rank
+
+
+def _transfer_to_standard_form(
+    M: np.array, n: int, k: int
+) -> Tuple[np.array, np.array, np.array, int]:
+    """
+    Puts the stabilizer matrix in its standardized form, as in section 4.1 of the thesis.
+
+    Args:
+        M: The stabilizier matrix, to be standardized.
+        n: Dimension of the code words.
+        k: Dimension of the message words.
+
+    Returns:
+        The standardized matrix.
+        The logical Xs.
+        The logical Zs.
+        The rank of the matrix.
+    """
+
+    # Performing the Gaussian elimination as in section 4.1
+    r: int = _GaussianElimination(M, 0, n - k, 0, n)
+    _ = _GaussianElimination(M, r, n - k, n + r, 2 * n)
+
+    # Get matrix sub-components, as per equation 4.3:
+    # A1 = M[0:r, r : (n - k)]
+    A2 = M[0:r, (n - k) : n]
+    # B = M[0:r, n : (n + r)]
+    C1 = M[0:r, (n + r) : (2 * n - k)]
+    C2 = M[0:r, (2 * n - k) : (2 * n)]
+    # D = M[r : (n - k), n : (n + r)]
+    E = M[r : (n - k), (2 * n - k) : (2 * n)]
+
+    X = np.concatenate(
+        [
+            np.zeros((k, r), dtype=np.int8),
+            E.T,
+            np.eye(k, dtype=np.int8),
+            np.mod(E.T @ C1.T + C2.T, 2),
+            np.zeros((k, n - r), np.int8),
+        ],
+        axis=1,
+    )
+
+    Z = np.concatenate(
+        [
+            np.zeros((k, n), dtype=np.int8),
+            A2.T,
+            np.zeros((k, n - k - r), dtype=np.int8),
+            np.eye(k, dtype=np.int8),
+        ],
+        axis=1,
+    )
+    return M, X, Z, r
+
+
+class StabilizerCode(object):
+    def __init__(self, group_generators: List[str], correctable_errors: List[str]):
+        n = len(group_generators[0])
+        k = n - len(group_generators)
+
+        # Build the matrix defined in section 3.4. Each row corresponds to one generator of the
+        # code, which is a vector of dimension n. The elements of the vectors are Pauli matrices
+        # encoded as I, X, Y, or Z. However, as described in the thesis, we encode the Pauli
+        # vector of 2*n Booleans.
+        M = np.zeros((n - k, 2 * n), np.int8)
+        for i, group_generator in enumerate(group_generators):
+            for j, c in enumerate(group_generator):
+                if c == 'X' or c == 'Y':
+                    M[i, j] = 1
+                elif c == 'Z' or c == 'Y':
+                    M[i, n + j] = 1
+
+        M, X, Z, r = _transfer_to_standard_form(M, n, k)
+
+        self.n: int = n
+        self.k: int = k
+        self.r: int = r
+        self.M: List[str] = _BuildByCode(M)
+        self.logical_Xs: List[str] = _BuildByCode(X)
+        self.logical_Zs: List[str] = _BuildByCode(Z)
+
+        self.syndromes_to_corrections = {}
```

Comment:
I'm surprised this doesn't need a type annotation.

Reply:
Tests seemed fine. Nevertheless, code would be clearer with an annotation so I added one.